pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java

```diff
@@ -1756,4 +1763,93 @@ public void testReplicatorProducerNotExceed() throws Exception {
 
         Assert.assertThrows(PulsarClientException.ProducerBusyException.class, () -> new MessageProducer(url2, dest2));
     }
+
+    @Test
+    public void testDiscontinuousMessages() throws Exception {
```

Comment:
Hi @codelipenghui 

> What kind of exceptions will happen to the producer? IMO, we should use infinite publish timeout, if the exception is not publish timeout, it should not be a retriable exception. We should stop the replication.

My description of the scene wasn't good, it is not an `exception` for sending messages. Now there are two scenarios:
- the schema of messages is loading. see Code-1 below
- something wrong when processing `replicateEntries`. see Code-2 below


#### Code-1
https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L160-L179

#### Code-2
https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L190-L192


I opened a comment list to easily track the context(^_^).

Reply:
Ok, if we can reach https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L190-L192, can we just close the replicator and try to restart the replication task?

The main purpose of the description of the PR is to avoid processing [4,5] before [1,2,3] is processed successfully. It should be a corner case with a very low incidence. If close the issue replicator can help with the problem, it will be more simpler for implementation.

Reply:
@codelipenghui 

> Ok, if we can reach https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L190-L192, can we just close the replicator and try to restart the replication task?

Background: Now the Replicator is having some trouble starting and closing:
- After closing the internal producer of Replicator, this Replicator will not work anymore, there has no mechanism to get it running again
  -  I will push another PR to fix it.
- The function `disconnect` maybe has a bug that makes the Replicator continues to work after the internal cursor is closed
  - When `startProducer` and `disconnect` run concurrently, the thread `disconnect` may not get the right producer. 
  - I will push another PR to fix it.

So there is not yet a stable way to restart Replicator(other than an `unload topic`).

---

### Summary

Now this PR tries to solve three problems in the same way: `Make the replicator processes messages only sequentially`
- scenario-1: race condition of `read entries` and `rewind`
- scenario-2: the task replicate messages is aborted by a loading schema
- scenario-3: the task replicate messages is aborted by an unknown error(we are talking about)

Since scenario-1 and scenario-2 are there, the solution `Make the replicator processes messages only sequentially` is needed, so we can do no additional processing on scenario-3.

Reply:
Will the current solution make the replicator not able to re-replicate data(reset the cursor of the replicator) to the remote cluster unless restart the replicator?

And will the solution be resolved by unloading the topic if the replicator runs into an exception?

Reply:
@codelipenghui 

> Will the current solution make the replicator not able to re-replicate data(reset the cursor of the replicator) to the remote cluster unless restart the replicator?

1. If someone moves Replicator's cursor to a larger pointer manually(such as `pulsar-admin topics resetCursor`), everything is ok. The logic of checking "whether the message is continuous or not" is `first_message_received <= Max(mark_deleted_pos, last_sent_pos) + 1`, since `resetCursor` will move the variable `markDeletePosition` to a larger value too, the continuous check will return `true`.
2. If someone moves Replicator's cursor to a smaller pointer manually(such as `rewind`), everything is ok.

> And will the solution be resolved by unloading the topic if the replicator runs into an exception?

I think so.

Reply:
> I mean if we don't apply this fix but try to "close the replicator -> unload the topic" if the replicator reaches https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L190-L192, is it able to fix the issue?

Yes, I think so.

Reply:
Ok, this way looks easier to understand and straightforward.
It's very similar to Pulsar connectors, like a connector to consume messages from a topic and publish them to another topic. I mean if want to introduce a principle to handle such issues, I think we don't want to add such logic to check if there are messages been skipped. This is why I think we should not introduce such logic here.

Reply:
@codelipenghui 

I see:
- for scenario-1: race condition of `read entries` and `rewind`
  - a separate PR or PIP will solve this problem
- for scenario-3: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java#L190-L192
  - restart Replicator to fix it. 

####  for scenario-2:  the task replicate messages is aborted by a loading schema

| time | `receive messages` | `async load schema` |
| --- | --- | --- |
| 1 | receive messages `1~3` | 
| 2 | async load schema if needed |
| 3 | receive messages `4~5`|
| 4 | send messages `4~5` to remote cluster | 
| 5 | | load schema success, rewind cursor to position `1`  |
| 6 | <strong>(Highlight)</strong>Receive messages `1~3` and will send fail by duplication check |

This scenario is hard to fix.  Because there is no clear marker of which in-flight reads need to be discarded after step-2. I think I should learn how the connector handle messages.

Reply:
> This scenario is hard to fix. Because there is no clear marker of which in-flight reads need to be discarded after step-2. I think I should learn how the connector handle messages.

I see. We can continue to handle these messages after loading schema is finished.

Reply:
@codelipenghui  I think this suggestion is good. Thanks. I will try to solve the issues described above with multiple PR

Reply:
> This scenario is hard to fix. Because there is no clear marker of which in-flight reads need to be discarded after step-2. I think I should learn how the connector handle messages.

But we will not get an error if the message is duplicated, right? The broker skips the duplicated sends and returns (-1, -1) as the message ID. We are safe to publish messages 1 to 5 again.